pyro/infer/mcmc/hmc.py

```diff
@@ -93,6 +108,35 @@ def _get_trace(self, z):
         trace_poutine(*self._args, **self._kwargs)
         return trace_poutine.trace
 
+    @staticmethod
+    def _iter_latent_nodes(trace):
+        for name, node in sorted(trace.iter_stochastic_nodes(), key=lambda x: x[0]):
+            if not (node["fn"].has_enumerate_support or isinstance(node["fn"], _Subsample)):
+                yield (name, node)
+
+    def _compute_trace_log_prob(self, model_trace):
+        if not self._has_enumerable_sites:
+            return model_trace.log_prob_sum()
+        if self.max_iarange_nesting == float("inf"):
+            raise ValueError("Finite value required for `max_iarange_nesting` when model "
+                             "has discrete (enumerable) sites.")
+        model_trace.compute_log_prob()
+        ordering = {name: frozenset(f for f in site["cond_indep_stack"])
+                    for name, site in model_trace.nodes.items()
+                    if site["type"] == "sample"}
+
+        log_probs = defaultdict(float)
+        for name, site in model_trace.nodes.items():
+            if site["type"] == "sample":
+                log_probs[ordering[name]] = log_probs[ordering[name]] + site["log_prob"]
+
+        log_prob_sum = 0.
+        for log_prob in log_probs.values():
+            while log_prob.dim() > self.max_iarange_nesting:
```

Comment:
nit: it would.be cheaper to reshape, something like:
```py
enum_dim = log_prob.dim() - self.max_iarange_nesting
if enum_dim > 0:
    log_prob = log_sum_exp(log_prob.reshape(-1, *log_prob.shape[enum_dim:]), dim=0)
```

Reply:
+1; will make the change.